src/linker/Linker/MarkingHelpers.cs

```diff
@@ -12,12 +11,30 @@ public MarkingHelpers (LinkContext context)
 			_context = context;
 		}
 
-		public void MarkExportedType (ExportedType type, ModuleDefinition module, in DependencyInfo reason)
+		public void MarkExportedType (ExportedType exportedType, ModuleDefinition module, in DependencyInfo reason)
 		{
-			if (!_context.Annotations.MarkProcessed (type, reason))
+			if (!_context.Annotations.MarkProcessed (exportedType, reason))
 				return;
+
+			if (reason.Kind == DependencyKind.ModuleOfExportedType) {
+				var facadeAssembly = (reason.Source as ModuleDefinition).Assembly;
+				TypeDefinition resolvedType = exportedType.Resolve ();
+				while (facadeAssembly.FullName != resolvedType.Module.Assembly.FullName) {
```

Comment:
Wondering if instead of trying to fix the forwarder chain we could get away with marking just the forwarder and then when we're writing out the output, we would "fix" it to point to a thing that exists (so if we're saving a forwarder, we would check whether the thing its pointing to is marked and follow that until we find the marked thing).

That way e.g. referencing System.Object from netstandard won't bring System.Runtime.dll.

Reply:
I think we do something almost like this at https://github.com/mono/linker/blob/main/src/linker/Linker.Steps/SweepStep.cs#L683 - it will fix up forwarders so that they point directly at a resolved type - but it won't handle cases where the type can't be resolved. So if you have a -> b -> c -> (unresolved) where a and c are marked forwarders, it won't rewrite a to point to c. But I don't think it will necessarily keep b either - it would be great to add a testcase for this.